cellmlmanip/units.py

```diff
@@ -622,3 +635,179 @@ def traverse(self, expr):
+        in a conversion factor) this will recursively traverse the expression tree and convert
+        at each level as needed. Hence if the operands of internal expressions are in dimensionally
+        consistent but not equal units, conversions will be applied as needed.
+
+        This method is suitable for use converting the RHS of assignment equations to the units
+        desired by the LHS, if the ``Eq`` expression is passed in as ``expr``, and ``to_units`` may
+        be given as ``None``.
+
+        The conversion strategy for each (sub-)expression depends on the operator:
+        - for relational operators, all operands are converted to the units of the first operand
+        - for Mul the operands can be in any units, and we convert the result if needed
+        - for Add all operands are converted to the desired units (or the units of the first operand
+          if no desired units are given)
+        - for Pow the exponent must be dimensionless while the operand can be in any units, and we
+          convert the result if needed
+        - for trig functions, exp, log, etc. the operands have to have dimensionless units
+        - for piecewise, the conditions must be dimensionless, and the pieces are set to the desired
+          units (or the units of the first piece)
+        - for derivatives, numbers, variables, etc. we just convert to the desired units
+
+        :param expr: the Sympy expression to convert
+        :param to_units: the desired units of the expression, or ``None`` if we don't care.
+        :returns: a Sympy expression in the desired units; the input ``expr`` if no conversion was needed.
+        """
+        new_expr, was_converted, actual_units = self._convert_expr_recursively(expr, to_units)
+        return new_expr
+
+    def set_lhs_units_from_rhs(self, equation):
+        """Set the units of the variable assigned to based on the expression assigned.
+
+        :param equation: a :class:`sympy.Eq` instance with a :class:`VariableDummy` on the left hand side
+        :returns: a new version of the equation, but with the RHS using consistent units throughout (with
+            conversions if needed) and the ``units`` attribute on the LHS updated to match
+        """
+        lhs = equation.lhs
+        assert isinstance(lhs, model.VariableDummy)
+        new_rhs, was_converted, rhs_units = self._convert_expr_recursively(equation.rhs, None)
+        lhs.units = rhs_units
+        if was_converted:
+            equation = sympy.Eq(lhs, new_rhs)
+        return equation
+
+    def _convert_expr_recursively(self, expr, to_units):
+        """Helper method for :meth:`convert_expr_recursively` which does the heavy lifting.
+
+        :returns: a tuple ``(new_expr, was_converted, actual_units)``
+        """
+        print('Trying to convert {} {} to {}'.format(type(expr), expr, to_units))
+        was_converted = False  # Tracks whether we needed a units conversion
+        new_args = []
```

Comment:
Can we define `new_args` only in the branches where we need it? Makes it more readable as well